frontend/src/metabase/common/components/EntityPicker/components/EntityPickerSearch/EntityPickerSearch.tsx

```diff
@@ -0,0 +1,134 @@
+import { t } from "ttag";
+
+import NoResults from "assets/img/no_results.svg";
+import EmptyState from "metabase/components/EmptyState";
+import { VariableSizeItemsVirtualizedList } from "metabase/components/VirtualizedList";
+import Search from "metabase/entities/search";
+import { useDebouncedEffect } from "metabase/hooks/use-debounced-effect";
+import { useDispatch } from "metabase/lib/redux";
+import { SearchLoadingSpinner } from "metabase/nav/components/search/SearchResults";
+import type { WrappedResult } from "metabase/search/types";
+import { Stack, Tabs, TextInput, Icon, Box, Flex } from "metabase/ui";
+import type {
+  SearchResult as SearchResultType,
+  SearchResults as SearchResultsType,
+} from "metabase-types/api";
+
+import type { TypeWithModel } from "../../types";
+
+import { EntityPickerSearchResult } from "./EntityPickerSearch.styled";
+
+const defaultSearchFilter = (results: SearchResultType[]) => results;
+
+export function EntityPickerSearchInput({
+  searchQuery,
+  setSearchQuery,
+  setSearchResults,
+  models,
+  searchFilter = defaultSearchFilter,
+}: {
+  searchQuery: string;
+  setSearchQuery: (query: string) => void;
+  setSearchResults: (results: SearchResultType[] | null) => void;
+  models: string[];
+  searchFilter?: (results: SearchResultType[]) => SearchResultType[];
+}) {
+  useDebouncedEffect(
+    () => {
+      if (searchQuery) {
+        Search.api
+          .list({ models, q: searchQuery })
+          .then((results: SearchResultsType) => {
+            if (results.data) {
+              const filteredResults = searchFilter(results.data);
+              setSearchResults(filteredResults);
+            } else {
+              setSearchResults(null);
+            }
+          });
+      } else {
+        setSearchResults(null);
+      }
+    },
+    200,
+    [searchQuery, models, searchFilter],
+  );
+
+  return (
+    <TextInput
+      type="search"
+      icon={<Icon name="search" size={16} />}
+      miw={400}
+      mr="2rem"
+      value={searchQuery}
+      onChange={e => setSearchQuery(e.target.value ?? "")}
+      placeholder={t`Search…`}
+    />
+  );
+}
+
+export const EntityPickerSearchResults = <TItem extends TypeWithModel>({
+  searchResults,
+  onItemSelect,
+  selectedItem,
+}: {
+  searchResults: SearchResultType[] | null;
+  onItemSelect: (item: TItem) => void;
+  selectedItem: TItem | null;
+}) => {
+  const dispatch = useDispatch();
+
+  if (!searchResults) {
+    return <SearchLoadingSpinner />;
+  }
+
+  return (
+    <Box p="lg" h="100%">
+      {searchResults.length > 0 ? (
+        <Stack h="100%">
+          <VariableSizeItemsVirtualizedList>
+            {searchResults?.map(item => (
+              <EntityPickerSearchResult
+                key={item.model + item.id}
+                result={Search.wrapEntity(item, dispatch)}
+                onClick={(item: WrappedResult) => {
+                  onItemSelect(item as unknown as TItem);
```

Comment:
why do we have to typecast?

Reply:
how much time you got? 😂

there's a couple type issues that are really tricky here. I'll grab some time with you to see if you have any better ideas than we came up with. these couple casts were by far the least bad option we came up with